core/primitives/src/block.rs

```diff
@@ -95,22 +93,23 @@ pub enum Block {
 
 pub fn genesis_chunks(
     state_roots: Vec<StateRoot>,
-    num_shards: NumShards,
+    shard_ids: &[ShardId],
     initial_gas_limit: Gas,
     genesis_height: BlockHeight,
     genesis_protocol_version: ProtocolVersion,
 ) -> Vec<ShardChunk> {
-    assert!(state_roots.len() == 1 || state_roots.len() == (num_shards as usize));
+    assert!(state_roots.len() == 1 || state_roots.len() == shard_ids.len());
     let mut rs = ReedSolomonWrapper::new(1, 2);
 
-    (0..num_shards)
-        .map(|i| {
+    shard_ids
+        .into_iter()
+        .map(|&shard_id| {
             let (encoded_chunk, _) = EncodedShardChunk::new(
                 CryptoHash::default(),
-                state_roots[i as usize % state_roots.len()],
+                state_roots[shard_id as usize % state_roots.len()],
```

Comment:
I don't understand the `% len` part.
Is it a valid input to have `shard_ids = [1,4,7]`? Then the assertion `state_roots.len() == shard_ids.len()` doesn't help, the shards will be initialized with the same state root.

Reply:
I also do not understand the `% len` part.  I left it as future work to figure out what is happening here. 

> Is it a valid input to have shard_ids = [1,4,7]?

No, in general, we have too many places in the code where we assume that shard ids are contiguous.  We are slowly trying to address all these issues.

Let me know if you any other questions on this PR.

Reply:
I don't think that `%len` is correct code. Suggestion below.

```suggestion
   let state_roots : Iter<StateRoot> = if state_roots.len() == shard_ids.len() {
        state_roots.iter()
    } else {
        std::iter::repeat(state_roots[0]).take(4)
    };
    shard_ids.iter().zip(state_roots).map(|(shard_id,state_root){
            let (encoded_chunk, _) = EncodedShardChunk::new(
                CryptoHash::default(),
                state_root,
                CryptoHash::default(),
                genesis_height,
                shard_id,
```

Reply:
Thanks for the suggestion!  I applied it with some minor edits.